File: relisoft-hr/server/controllers/internalMobilityController.js
import InternalJobPosting from '../models/InternalJobPosting.js';
import InternalApplication from '../models/InternalApplication.js';
import Employee from '../models/Employee.js';

export const createPosting = async (req, res) => {
  try {
    const employee = await Employee.findOne({ userId: req.user.id });
    if (!employee) return res.status(404).json({ success: false, message: 'Employee not found' });

    const posting = await InternalJobPosting.create({ ...req.body, postedBy: employee._id });
    res.status(201).json({ success: true, data: posting });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const getPostings = async (req, res) => {
  try {
    const { department, status } = req.query;
    const query = {};
    if (!status) query.status = 'Open';
    else query.status = status;
    if (department) query.department = department;

    const postings = await InternalJobPosting.find(query).populate('postedBy', 'firstName lastName').sort('-createdAt');
    res.status(200).json({ success: true, data: postings });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const getPosting = async (req, res) => {
  try {
    const posting = await InternalJobPosting.findById(req.params.id).populate('postedBy', 'firstName lastName');
    if (!posting) return res.status(404).json({ success: false, message: 'Posting not found' });
    res.status(200).json({ success: true, data: posting });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const updatePosting = async (req, res) => {
  try {
    const posting = await InternalJobPosting.findByIdAndUpdate(req.params.id, req.body, { new: true, runValidators: true });
    if (!posting) return res.status(404).json({ success: false, message: 'Posting not found' });
    res.status(200).json({ success: true, data: posting });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const deletePosting = async (req, res) => {
  try {
    const posting = await InternalJobPosting.findByIdAndDelete(req.params.id);
    if (!posting) return res.status(404).json({ success: false, message: 'Posting not found' });
    res.status(200).json({ success: true, data: {} });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const createApplication = async (req, res) => {
  try {
    const employee = await Employee.findOne({ userId: req.user.id });
    if (!employee) return res.status(404).json({ success: false, message: 'Employee not found' });

    const existing = await InternalApplication.findOne({ posting: req.body.posting, employee: employee._id });
    if (existing) return res.status(400).json({ success: false, message: 'You have already applied to this posting' });

    const application = await InternalApplication.create({ ...req.body, employee: employee._id });
    res.status(201).json({ success: true, data: application });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const getApplications = async (req, res) => {
  try {
    const { employeeId } = req.params;
    const applications = await InternalApplication.find({ employee: employeeId })
      .populate('posting', 'title department location')
      .populate('employee', 'firstName lastName employeeId')
      .populate('approvedBy', 'firstName lastName')
      .sort('-createdAt');
    res.status(200).json({ success: true, data: applications });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

export const updateApplicationStatus = async (req, res) => {
  try {
    const { status } = req.body;
    const application = await InternalApplication.findByIdAndUpdate(
      req.params.id,
      { status },
      { new: true, runValidators: true }
    );
    if (!application) return res.status(404).json({ success: false, message: 'Application not found' });
    res.status(200).json({ success: true, data: application });
  } catch (error) {
    res.status(500).json({ success: false, message: error.message });
  }
};

File: relisoft-hr/server/Controllers/InternalMobilityController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using RelisoftHR.Data;
using RelisoftHR.DTOs;
using RelisoftHR.Models;
using RelisoftHR.Services;

namespace RelisoftHR.Controllers;

[ApiController]
[Route("api/internal-mobility")]
[Authorize]
public class InternalMobilityController : ControllerBase
{
    private readonly AppDbContext _db;
    public InternalMobilityController(AppDbContext db) => _db = db;

    private int GetUserId()
    {
        var claim = User.FindFirst(System.Security.Claims.ClaimTypes.NameIdentifier)?.Value;
        return claim != null ? int.Parse(claim) : 0;
    }

    [HttpGet("jobs")]
    public async Task<ActionResult> GetJobs()
    {
        var list = await _db.InternalJobPostings
            .Where(j => j.IsActive)
            .OrderByDescending(j => j.PostingDate)
            .ToListAsync();
        return Ok(list);
    }

    [HttpPost("jobs")]
    public async Task<ActionResult> CreateJob([FromBody] InternalJobPostingRequest req)
    {
        var posting = new InternalJobPosting
        {
            Title = req.Title, Description = req.Description, Requirements = req.Requirements,
            Department = req.Department, Location = req.Location, PostingDate = req.PostingDate,
            ClosingDate = req.ClosingDate, CreatedById = GetUserId(), CreatedOn = DateTime.UtcNow, IsActive = true
        };
        _db.InternalJobPostings.Add(posting);
        await _db.SaveChangesAsync();
        return Ok(posting);
    }

    [HttpPut("jobs/{id}")]
    public async Task<ActionResult> UpdateJob(int id, [FromBody] InternalJobPostingRequest req)
    {
        var j = await _db.InternalJobPostings.FindAsync(id);
        if (j == null) return NotFound();
        HttpConcurrency.RequireIfMatch(Request, _db, j);
        j.Title = req.Title;
        j.Description = req.Description;
        j.Requirements = req.Requirements;
        j.Department = req.Department;
        j.Location = req.Location;
        j.ClosingDate = req.ClosingDate;
        await _db.SaveChangesAsync();
        HttpConcurrency.SetETag(Response, j.RowVersion);
        return Ok(j);
    }

    [HttpPost("jobs/{id}/close")]
    public async Task<ActionResult> CloseJob(int id)
    {
        var j = await _db.InternalJobPostings.FindAsync(id);
        if (j == null) return NotFound();
        j.IsActive = false;
        await _db.SaveChangesAsync();
        return Ok(new { message = "Closed" });
    }

    [HttpGet("applications/my")]
    public async Task<ActionResult> GetMyApplications()
    {
        var empId = GetUserId();
        var list = await _db.InternalJobApplications
            .Include(a => a.JobPosting)
            .Where(a => a.EmployeeId == empId)
            .OrderByDescending(a => a.CreatedOn)
            .ToListAsync();
        return Ok(list);
    }

    [HttpPost("jobs/{jobId}/apply")]
    public async Task<ActionResult> Apply(int jobId, [FromBody] InternalJobApplicationRequest req)
    {
        var job = await _db.InternalJobPostings.FindAsync(jobId);
        if (job == null) return NotFound();
        if (!job.IsActive || job.ClosingDate < DateTime.UtcNow)
            return BadRequest(new { message = "This job posting is closed" });
        var employeeId = GetUserId();
        if (await _db.InternalJobApplications.AnyAsync(a =>
            a.JobPostingId == jobId && a.EmployeeId == employeeId))
            return Conflict(new { message = "You have already applied for this job" });
        var application = new InternalJobApplication
        {
            JobPostingId = jobId, EmployeeId = employeeId, CoverNote = req.CoverNote,
            Status = "Applied", CreatedOn = DateTime.UtcNow
        };
        _db.InternalJobApplications.Add(application);
        await _db.SaveChangesAsync();
        return Ok(application);
    }

    [HttpGet("applications")]
    public async Task<ActionResult> GetAllApplications()
    {
        var empId = GetUserId();
        var list = await _db.InternalJobApplications
            .Include(a => a.Employee)
            .Include(a => a.JobPosting)
            .Where(a => a.JobPosting != null && a.JobPosting.CreatedById == empId)
            .OrderByDescending(a => a.CreatedOn)
            .ToListAsync();
        return Ok(list);
    }

    [HttpPost("applications/{id}/shortlist")]
    public async Task<ActionResult> Shortlist(int id)
    {
        var app = await _db.InternalJobApplications.FindAsync(id);
        if (app == null) return NotFound();
        app.Status = "Shortlisted";
        await _db.SaveChangesAsync();
        return Ok(app);
    }

    [HttpPost("applications/{id}/reject")]
    public async Task<ActionResult> RejectApplication(int id)
    {
        var app = await _db.InternalJobApplications.FindAsync(id);
        if (app == null) return NotFound();
        app.Status = "Rejected";
        await _db.SaveChangesAsync();
        return Ok(app);
    }
}
